sbi/inference/snpe/snpe_c.py

```diff
@@ -231,5 +370,354 @@ def _log_prob_proposal_posterior(
+
+        means_pp = _batched_mv(covariances_pp, summed_cov_m_prod_rep)
+
+        return means_pp
+
+    @staticmethod
+    def _logits_proposal_posterior(
+        means_pp: Tensor,
+        precisions_pp: Tensor,
+        covariances_pp: Tensor,
+        logits_p: Tensor,
+        means_p: Tensor,
+        precisions_p: Tensor,
+        logits_d: Tensor,
+        means_d: Tensor,
+        precisions_d: Tensor,
+    ):
+        """
+        Return the component weights (i.e. logits) of the proposal posterior.
+
+        Args:
+            means_pp: Means of the proposal posterior.
+            precisions_pp: Precision matrices of the proposal posterior.
+            covariances_pp: Covariance matrices of the proposal posterior.
+            logits_p: Component weights (i.e. logits) of the proposal distribution.
+            means_p: Means of the proposal distribution.
+            precisions_p: Precision matrices of the proposal distribution.
+            logits_d: Component weights (i.e. logits) of the density estimator.
+            means_d: Means of the density estimator.
+            precisions_d: Precision matrices of the density estimator.
+
+        Returns: Component weights of the proposal posterior. L*K terms.
+        """
+
+        num_comps_p = precisions_p.shape[1]
+        num_comps_d = precisions_d.shape[1]
+
+        # Compute alpha_i * beta_j
+        logits_p_rep = logits_p.repeat_interleave(num_comps_d, dim=1)
+        logits_d_rep = logits_d.repeat(1, num_comps_p)
+        logit_factors = logits_p_rep + logits_d_rep
+
+        # Compute sqrt(det()/(det()*det()))
+        covariances_p = torch.inverse(precisions_p)
+        covariances_d = torch.inverse(precisions_d)
+
+        logdet_covariances_pp = torch.logdet(covariances_pp)
+        logdet_covariances_p = torch.logdet(covariances_p)
+        logdet_covariances_d = torch.logdet(covariances_d)
+
+        # Repeat the proposal and density estimator terms such that there are LK terms.
+        # Same trick as has been used above.
+        logdet_covariances_p_rep = logdet_covariances_p.repeat_interleave(
+            num_comps_d, dim=1
+        )
+        logdet_covariances_d_rep = logdet_covariances_d.repeat(1, num_comps_p)
+
+        sqrt_det_ratio = 0.5 * (
+            logdet_covariances_pp
+            - (logdet_covariances_p_rep + logdet_covariances_d_rep)
+        )
+
+        # Compute for proposal, density estimator, and proposal posterior:
+        # mu_i.T * P_i * mu_i
+        exponent_p = _batched_vmv(precisions_p, means_p)
+        exponent_d = _batched_vmv(precisions_d, means_d)
+        exponent_pp = _batched_vmv(precisions_pp, means_pp)
+
+        # Extend proposal and density estimator exponents to get LK terms.
+        exponent_p_rep = exponent_p.repeat_interleave(num_comps_d, dim=1)
+        exponent_d_rep = exponent_d.repeat(1, num_comps_p)
+        exponent = -0.5 * (exponent_p_rep + exponent_d_rep - exponent_pp)
```

Comment:
why isn't the `exponent_pp` repeated as well? doesn't it have L*K components?

Reply:
`exponent_pp` is computed from `means_pp` and `precision_pp`. Both of these tensors have already been repeated before. So `exponent_pp` directly has LK components.